cleverhans/attacks.py

```diff
@@ -486,6 +486,102 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
         return True
 
 
+class CarliniWagnerL2(Attack):
+    """
+    This attack was originally proposed by Carlini and Wagner. It is an
+    iterative attack that finds adversarial examples on many defenses that
+    are robust to other attacks.
+    Paper link: https://arxiv.org/abs/1608.04644
+
+    """
+    def __init__(self, model, back='tf', sess=None):
+        super(CarliniWagnerL2, self).__init__(model, back, sess)
+
+        if self.back == 'th':
+            raise NotImplementedError('Theano version not implemented.')
+        self.attack_objects = {}
+
+    def generate(self, x, y=None, nb_classes=10,
+                 batch_size=1, confidence=0,
+                 targeted=True, learning_rate=1e-3,
+                 binary_search_steps=10, max_iterations=100,
+                 abort_early=True, initial_const=1e-2,
+                 clip_min=0, clip_max=1):
+
+        import tensorflow as tf
+        from .attacks_tf import CarliniWagnerL2 as CWL2
+
+        attack = CWL2(self.sess, self.model, batch_size, confidence, targeted,
+                      learning_rate, binary_search_steps, max_iterations,
+                      abort_early, initial_const, clip_min, clip_max,
+                      nb_classes, x.get_shape().as_list()[1:])
+
+        def cw_wrap(x_val, y_val):
+            return np.array(attack.attack(x_val, y_val), dtype=np.float32)
+
+        wrap = tf.py_func(cw_wrap, [x, y], tf.float32)
+        return wrap
+
+    def generate_np(self, x_val, y_val=None, nb_classes=10,
```

Comment:
Same comment about the arguments being parsed outside of the `parse_params` method.